[HOP][inductor] Support pytrees as associative_scan input

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122137
* #119430

This allows `associative_scan` to take an arbitrary pytree of tensors,
which is flattened to their leaves before calling the `associative_scan`
higher order operator.

I also add support in inductor to generate code for scanning over sequences
of tensors.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -889,27 +889,24 @@ def arg_extractor(input, dim, combine_fn):
 
         input, dim, combine_fn = arg_extractor(*args, **kwargs)
 
-        assert isinstance(
```
why?

peterbell10 replied:
The use of `functools.partial` to wrap the `combine_fn` resulted in this becoming `FunctoolsPartialVariable`.

lezcano commented on torch/_inductor/subgraph_lowering.py:
```diff
@@ -87,7 +88,7 @@ def placeholder(self, idx):
     def output(self, *args):
         return self.tracer.create_node(
-            "output", "output", tuple(self.tracer.create_arg(a) for a in args), {}
+            "output", "output", (tuple(self.tracer.create_arg(a) for a in args),), {}
```
why the wrapping and unwrapping in line 66?

peterbell10 replied:
Apparently the fx output operator doesn't actually support multiple arguments, it will silently ignore anything but `arg[0]`. So you need to explicitly pass a tuple of results.

lezcano reviewed (approved):
Most of the heavy lifting is done in `_higher_order_ops`. I'll let @zou3519 review that. The rest of the code seems like mechanical changes. That part LGTM

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Use speculate_subgraph and reorder args on "[HOP][inductor] Support pytrees as associative_scan input"


This allows `associative_scan` to take an arbitrary pytree of tensors,
which is flattened to their leaves before calling the `associative_scan`
higher order operator.

I also add support in inductor to generate code for scanning over sequences
of tensors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 committed:
Update on "[HOP][inductor] Support pytrees as associative_scan input"


This allows `associative_scan` to take an arbitrary pytree of tensors,
which is flattened to their leaves before calling the `associative_scan`
higher order operator.

I also add support in inductor to generate code for scanning over sequences
of tensors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 committed:
Update on "[HOP][inductor] Support pytrees as associative_scan input"


This allows `associative_scan` to take an arbitrary pytree of tensors,
which is flattened to their leaves before calling the `associative_scan`
higher order operator.

I also add support in inductor to generate code for scanning over sequences
of tensors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 committed:
Lint

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update on "[HOP][inductor] Support pytrees as associative_scan input"


This allows `associative_scan` to take an arbitrary pytree of tensors,
which is flattened to their leaves before calling the `associative_scan`
higher order operator.

I also add support in inductor to generate code for scanning over sequences
of tensors.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
